core/src/Serial/Kokkos_Serial.cpp

```diff
@@ -59,6 +68,16 @@ void SerialInternal::finalize() {
   }
 
   m_is_initialized = false;
+
+  // guard erasing from all_instances
+  {
+    std::scoped_lock lock(all_instances_mutex);
+    auto it = std::find(all_instances.begin(), all_instances.end(), this);
+    if (it == all_instances.end())
+      Kokkos::abort(
+          "Execution space instance to be removed couldn't be found!");
+    all_instances.erase(it);
```

Comment:
This is interesting it means we could now diagnose at finalize that some exec space instance has not been destroyed yet.
I am not suggesting you implement it now, just one thought.